rfd/0001-general-design.md

```diff
@@ -0,0 +1,433 @@
+---
+authors: Anton Repin (github.com/xdire)
+state: wip
+---
+
+# RFD 1 - General Design
+
+## What
+
+This RFD Details the aspects of implementation of TCP load balancer
+up the following minimal requirements:
+
+### Library
+- A least connections forwarder implementation that tracks the number of connections per upstream.
+- A per-client connection rate limiter implementation that tracks the number of client connections.
+- Health checking for unhealthy upstreams. Should not be eligible to receive connections until upstream determined to be healthy.
```

Comment:
Thanks for clarifying - please include this within the document.

Reply:
The document has been updated. I am still fixing typos and re-reading before re-requesting the review; however, you can parse through the updated file right away.

Reply:
Thanks!

Reply:
Perhaps my eyes are missing it - but I can't see a succinct explanation similar to this within the document - it might've been lost in the rest of the details. Maybe it's worth explicitly explaining this flow. The flow you've described sounds good though - it'd be useful for you to also briefly touch on how this could be improved in future to account for more complex degradations (e.g mitigating flappy backends by requiring multiple check passes before re-adding to the healthy pool)

Reply:
Thanks for the hints; I will update it with an additional diagram. Currently found the following missing from the document:
- Least connection selector detailed flow (still slice reference explained, not sure if needed diagram for quite a simple loop)
- Health check flow diagram and actual flow explanation (or diagram)

I will try to add in 30m.